arelle/XPathParser.py

```diff
@@ -1,227 +1,343 @@
 '''
 See COPYRIGHT.md for copyright information.
 '''
+from __future__ import annotations
+
 import sys
-from numbers import Number
+import time
+import traceback
+from collections.abc import Callable, Iterable, Sequence
+from decimal import Decimal
+from typing import Any, TYPE_CHECKING, Union
+from xml.dom import minidom
 
-from arelle import PythonUtil # define 2.x or 3.x string types (only needed when running as unit test from __main__
-from arelle.PluginManager import pluginClassMethods
 from pyparsing import (
-    Word, Keyword, alphas, ParseException, ParseSyntaxException,Literal, CaselessLiteral,Combine, Optional, nums, Forward, Group, ZeroOrMore, StringEnd, alphanums,ParserElement, quotedString,
-    delimitedList, Suppress, Regex
+    CaselessLiteral,
+    Combine,
+    Forward,
+    Group,
+    Keyword,
+    Literal,
+    Opt,
+    ParseBaseException,
+    ParseException,
+    ParseResults,
+    ParseSyntaxException,
+    ParserElement,
+    Regex,
+    StringEnd,
+    Suppress,
+    Word,
+    ZeroOrMore,
+    alphanums,
+    alphas,
+    delimited_list,
+    nums,
+    quoted_string,
 )
+
+from arelle import ModelValue, XbrlConst, XmlUtil
 from arelle.Locale import format_string
-import time, xml.dom, traceback
-from decimal import Decimal
-from arelle import (XmlUtil, ModelValue, XbrlConst)
-FunctionIxt = None
+from arelle.PluginManager import pluginClassMethods
+
+if TYPE_CHECKING:
+    from arelle.ModelFormulaObject import ModelFormulaResource
+    from arelle.ModelXbrl import ModelXbrl
+    from arelle.ModelManager import ModelManager
+    from arelle.ModelObject import ModelObject
+    from arelle.ModelValue import QName
+    from arelle.XPathContext import XPathContext, XPathException
+    from arelle.typing import TypeGetText
+
+    _: TypeGetText  # Handle gettext
+
+    FormulaToken = Union[
+        float,
+        int,
+        str,
+        Decimal,
+        'Expr',
+        'OpDef',
+        'OperationDef',
+        'ProgHeader',
+        'QNameDef',
+        'RangeDecl',
+        'VariableRef',
+    ]
+
+    RecursiveFormulaTokens = Sequence[Union[FormulaToken, 'RecursiveFormulaTokens']]
+
+    ExpressionStack = list[FormulaToken]
+
+ixtNamespaceFunctions: dict[str, dict[str, Callable[[str], str]]] | None = None
 
 
 # Debugging flag can be set to either "debug_flag=True" or "debug_flag=False"
-debug_flag=True
+debug_flag = True
+
+exprStack: ExpressionStack = []
+xmlElement: ModelObject | None = None
+modelXbrl: ModelXbrl | None = None
+pluginCustomFunctions: dict[QName, Callable[[XPathContext, OperationDef, ModelObject, list[list[FormulaToken]]], Any]] | None = None
```

Comment:
Out of curiosity, is the value of the dict actually used somewhere?  It seemed like `in` is the only operator used, so perhaps it could just be a set.

Anyway, I noticed that the `list[list[FormulaToken]]` here seems to mismatch `MATH_FUNCTION_ARGS`, and I was trying to understand if `MATH_FUNCTION_ARGS` is actually the correct type, or whether those functions should actually be verifying the types of the arguments.

Reply:
> Out of curiosity, is the value of the dict actually used somewhere? It seemed like `in` is the only operator used, so perhaps it could just be a set.

Not here. I kept the signature so I'd have it as reference while I'm working on the rest of the formula types.

> Anyway, I noticed that the `list[list[FormulaToken]]` here seems to mismatch `MATH_FUNCTION_ARGS`, and I was trying to understand if `MATH_FUNCTION_ARGS` is actually the correct type, or whether those functions should actually be verifying the types of the arguments.

Good catch! The numeric args func handles `list[list[FormulaToken]]` -> `Number`. Updated!